build-tools/packages/build-cli/test/commands/typetests.test.ts

```diff
@@ -2,11 +2,173 @@
+				const pkg = packageWithTypeValidation();
+				const expected = packageWithTypeValidation();
+				expect(pkg.typeValidation?.disabled).is.false;
+				updateTypeTestDependency(pkg, VersionOptions.ClearIfDisabled);
+				expect(pkg).to.deep.equal(expected);
+			});
+
+			it("removes previous test package dependency when type tests are disabled", () => {
+				const pkg = packageWithTypeValidation(/* enabled */ false);
+				expect(pkg.devDependencies?.["test-package-previous"]).to.exist;
+				updateTypeTestDependency(pkg, VersionOptions.ClearIfDisabled);
+				expect(pkg.devDependencies?.["test-package-previous"]).to.not.exist;
+			});
+		});
+
+		describe("VersionOptions.Previous", () => {
+			it("sets previous version", () => {
+				const pkg = packageWithTypeValidation();
+				updateTypeTestDependency(pkg, VersionOptions.Previous);
+				expect(pkg.devDependencies?.["test-package-previous"]).to.equal(
+					"npm:test-package@5.0.0",
+				);
+			});
+
+			it("sets previous version even without typeValidation node", () => {
+				const pkg = packageMinimal();
+				updateTypeTestDependency(pkg, VersionOptions.Previous);
+				expect(pkg.devDependencies?.["test-package-previous"]).to.equal(
+					"npm:test-package@5.0.0",
+				);
+			});
+
+			it("removes previous test package dependency when type tests are disabled", () => {
+				const pkg = packageWithTypeValidation(/* enabled */ false);
+				expect(pkg.devDependencies?.["test-package-previous"]).to.exist;
+				updateTypeTestDependency(pkg, VersionOptions.Previous);
+				expect(pkg.devDependencies?.["test-package-previous"]).to.not.exist;
+			});
+		});
+	});
+
+	describe("resetBrokenTests", () => {
+		it("empty", () => {
+			const pkgJson: { typeValidation?: ITypeValidationConfig } = {
+				typeValidation: {
+					broken: {},
+				},
+			};
+			resetBrokenTests(pkgJson);
+			assert.deepEqual(pkgJson, { typeValidation: { broken: {} } });
+		});
+
+		it("minimal", () => {
+			const pkgJson: { typeValidation?: ITypeValidationConfig } = {
+				typeValidation: {
+					broken: {
+						"broken-api": {
+							backCompat: false,
+							forwardCompat: false,
+						},
+					},
+				},
+			};
+			resetBrokenTests(pkgJson);
+			assert.deepEqual(pkgJson, { typeValidation: { broken: {} } });
+		});
+
+		it("ignores packages with no typeValidation node", () => {
+			const pkg = packageMinimal();
+			const expected = packageMinimal();
+
+			expect(pkg.typeValidation).to.not.exist;
+			resetBrokenTests(pkg);
+			expect(pkg).to.deep.equal(expected);
+		});
+
+		it("resets package with type tests", () => {
+			const pkg = packageWithTypeValidation();
+			resetBrokenTests(pkg);
+			expect(pkg.typeValidation?.broken).to.deep.equal({});
+		});
+
+		it("resets even if type validation is disabled", () => {
+			const pkg = packageWithTypeValidation(/* enabled */ false);
+			resetBrokenTests(pkg);
+			expect(pkg.typeValidation?.broken).to.deep.equal({});
+		});
```

Comment:
Tests have a cost (implemetaintion, maintaince, review, updating when the implementation changes etc). I feel like this test suite has a complexity much higher than the implementation.

One of the reasons I suggested making the input type so much more constrained is that now we can trust the compiler to ensure this function prabably doesn't screw up the fields it doesn't know about. This isn't guarenteed, but it should be enough for the level of confidence we want here.


Thus I think these last three tests really wouldn't be worth writing (they have more coupling than the other more targeted ones since they involve your full package information). I'd view these as more integration level tests, and with good unit testing and precise typing, we shouldn't need them. That said they seem reasonable and you already made them and I already reviewed them, so keeping them is also ok.

Reply:
> I'd view these as more integration level tests, and with good unit testing and precise typing, we shouldn't need them.

Good point about integration vs. unit tests. I get nervous I'm not considering all the relevant cases, Need more experience.

> That said they seem reasonable and you already made them and I already reviewed them, so keeping them is also ok.

Thanks, I'll leave them and get this merged, but I think I get what you mean.